pkg/cache/internal/informers.go

```diff
@@ -92,6 +107,25 @@ type tracker struct {
 	Metadata     map[schema.GroupVersionKind]*Cache
 }
 
+// GetOptions provides configuration to customize the behavior when
+// getting an informer.
+type GetOptions struct {
```

Comment:
(Just trying to understand how it fits together)

How is this option related to allowing to stop informers?

Is it that in your scenario the informers are started without waiting for the cache and if they don't come up you call the stop func?

Reply:
This bit is more about the utility of and usage patterns for stoppable informers.

In the original code, when a user requests an informer, controller-runtime blocks until that informer's cache is populated. This makes sense for imperative requests, like retrieving a specific object, since the cache needs to be read from for the request to be satisfied. This is not strictly necessary for registering a listener with an informer, since that listener will simply be dormant until events are fed to it.

Gatekeeper has a two-controller model for dynamic informers. There is the main controller and the sub controller. The main controller knows how to derive what GVKs the sub controller should be receiving events for. It then registers a listener with that GVK's associated informer (constructing it, if necessary). This listener passes events to the sub controller's work queue.

I should note I'm being a bit hand-wavy here as Gatekeeper actually uses a centralized structure to keep track of "interested" controllers (essentially a home-grown version of the reference counting solution controller-runtime is considering). We call it the "watch manager".

By allowing this centralized reference counter the ability to get the informer without blocking on it being ready, the watch manager can add/remove listeners without waiting for the informer to bootstrap (it's preferable to have initializing a watch be a synchronous operation so that the managing controller's reconciler can return whether or not a retry is warranted). This reduces the amount of time that main controllers can block each other.

Of course, sub controllers will eventually need to run a `Get()` request in order to run their reconcile loop, and that will be blocked on bootstrapping but:

1) if the sub controller is receiving events, that means the initial `List()` operation of the informer has been completed and therefore bootstrapping is likely also complete, or nearly complete.

2) there are fewer goroutines involved. A blocked sub controller only blocks itself. Main controllers can block each other.

TL;DR: for efficiency's sake, it'd be nice if adding/removing listeners didn't block on a bootstrapping process which they are not dependent on.

Reply:
That makes sense but seems unrelated to the rest of this PR, can this be it's own change? We seem to be unclear about what the best name is for this but other than that this bit seems uncontroversial

Reply:
I have no objections to splitting non-blocking-get into its own PR. I'll wait a day or so to open up a new PR to give people time to object.

Reply:
Split into https://github.com/kubernetes-sigs/controller-runtime/pull/2371